src/select/Select.ts

```diff
@@ -0,0 +1,411 @@
+	}
+
+	// custom select events
+	private _openSelect() {
+		this._open = true;
+		this._ignoreBlur = false;
+		this._focusedIndex = this._focusedIndex || 0;
+		this.invalidate();
+	}
+	private _closeSelect() {
+		this._open = false;
+		this.invalidate();
+	}
+
+	private _onTriggerClick(event: MouseEvent) {
+		this.properties.onClick && this.properties.onClick(event);
+
+		this._open ? this._closeSelect() : this._openSelect();
+	}
+
+	private _onTriggerBlur(event: FocusEvent) {
+		if (!this._ignoreBlur) {
+			this.properties.onBlur && this.properties.onBlur(event);
+			this._closeSelect();
+		}
+	}
+
+	private _onOptionMouseDown() {
+		this._ignoreBlur = true;
+	}
+
+	private _onOptionClick(event: MouseEvent) {
+		const {
+			options = [],
+			onChange,
+			onClick
+		} = this.properties;
+
+		onClick && onClick(event);
+
+		// find parent option node, and get index
+		let optionNode = <HTMLElement> event.target;
+		while (!optionNode.hasAttribute('data-dojo-index') && optionNode.parentElement) {
+			optionNode = optionNode.parentElement;
+		}
+
+		const index = optionNode.getAttribute('data-dojo-index');
+		if (index) {
+			const option = options[parseInt(index, 10)];
+
+			if (option && !option.disabled) {
+				this._focusedIndex = parseInt(index, 10);
+				onChange && onChange(option);
+			}
+		}
+	}
+
+	private _onListboxKeyDown(event: KeyboardEvent) {
+		this.properties.onKeyDown && this.properties.onKeyDown(event);
+
+		const {
+			options = [],
+			multiple,
+			onChange
+		} = this.properties;
+		const { _focusedIndex } = this;
+
+		switch (event.which) {
+			case keys.enter:
+				!options[_focusedIndex].disabled && onChange && onChange(options[_focusedIndex]);
+				break;
+			case keys.space:
+				!options[_focusedIndex].disabled && onChange && onChange(options[_focusedIndex]);
+				break;
+			case keys.escape:
+				this._closeSelect();
+				break;
+			case keys.down:
+				event.preventDefault();
+				if (this._open || multiple) {
+					this._focusedIndex = (_focusedIndex + 1) % options.length;
+				} else {
+					this._openSelect();
+				}
+				this.invalidate();
+				break;
+			case keys.up:
+				event.preventDefault();
+				this._focusedIndex = (_focusedIndex - 1 + options.length) % options.length;
+				this.invalidate();
+				break;
+			case keys.home:
+				this._focusedIndex = 0;
+				this.invalidate();
+				break;
+			case keys.end:
+				this._focusedIndex = options.length - 1;
+				this.invalidate();
+				break;
+		}
+	}
+
+	private _renderCustomOptions(): DNode[] {
+		const {
+			multiple,
+			options = [],
+			renderOption,
+			value
+		} = this.properties;
+
+		const optionNodes = [];
+		let optionNode, option;
+		for (let i = 0; i < options.length; i++) {
+			option = options[i];
+			option.id = option.id || uuid() + '';
+			option.selected = multiple ? option.selected : value === option.value;
+			optionNode = renderOption ? renderOption(option) : option.label;
+
+			optionNodes.push(v('div', {
```

Comment:
Yeah, in the end it didn't seem necessary, and I didn't really want to split out custom option nodes into a separate widget. I'm open to hearing reasons to do so, though